nemo/collections/asr/parts/rnnt_beam_decoding.py

```diff
@@ -0,0 +1,680 @@
+                    therefore it is required to use larger beam sizes to achieve the same (or close to the same)
+                    decoding accuracy as TSD.
+
+                    For a given decoding accuracy, it is possible to attain faster decoding via ALSD than TSD.
+
+            score_norm: bool, whether to normalize the scores of the log probabilities.
+
+            return_best_hypothesis: bool, decides whether to return a single hypothesis (the best out of N),
+                or return all N hypothesis (sorted with best score first). The container class changes based
+                this flag -
+                When set to True (default), returns a single Hypothesis.
+                When set to False, returns a NBestHypotheses container, which contains a list of Hypothesis.
+
+            # The following arguments are specific to the chosen `search_type`
+
+            tsd_max_sym_exp_per_step: Used for `search_type=tsd`. The maximum symmetric expansions allowed
+                per timestep during beam search. Larger values should be used to attempt decoding of longer
+                sequences, but this in turn increases execution time and memory usage.
+
+            alsd_max_target_len: Used for `search_type=alsd`. The maximum expected target sequence length
+                during beam search. Larger values allow decoding of longer sequences at the expense of
+                execution time and memory.
+
+            # The following two flags are placeholders and unused until `nsc` implementation is stabilized.
+
+            nsc_max_timesteps_expansion: Unused int.
+
+            nsc_prefix_alpha: Unused int.
+        """
+        self.decoder = decoder_model
+        self.joint = joint_model
+
+        self.blank = decoder_model.blank_idx
+        self.vocab_size = decoder_model.vocab_size
+        self.search_type = search_type
+        self.return_best_hypothesis = return_best_hypothesis
+
+        if beam_size < 1:
+            raise ValueError("Beam search size cannot be less than 1!")
+
+        self.beam_size = beam_size
+        self.score_norm = score_norm
+
+        if self.beam_size == 1:
+            self.search_algorithm = self.greedy_search
+        elif search_type == "default":
+            self.search_algorithm = self.default_beam_search
+        elif search_type == "tsd":
+            self.search_algorithm = self.time_sync_decoding
+        elif search_type == "alsd":
+            self.search_algorithm = self.align_length_sync_decoding
+        elif search_type == "nsc":
+            raise NotImplementedError("`nsc` (Constrained Beam Search) has not been implemented.")
+            # self.search_algorithm = self.nsc_beam_search
+        else:
+            raise NotImplementedError(
+                f"The search type ({search_type}) supplied is not supported!\n"
+                f"Please use one of : (default, tsd, alsd, nsc)"
```

Comment:
Or greedy?

Reply:
`greedy` is a special case when beam size is 1. It auto switches, and isnt an explicit case.